src/main/java/com/mariamkatamashvlii/gym/service/implementation/TrainerServiceImpl.java

```diff
@@ -0,0 +1,194 @@
+import com.mariamkatamashvlii.gym.dto.trainerDto.UpdateRequestDTO;
+import com.mariamkatamashvlii.gym.dto.trainerDto.UpdateResponseDTO;
+import com.mariamkatamashvlii.gym.dto.trainingDto.TrainerTrainingsRequestDTO;
+import com.mariamkatamashvlii.gym.dto.trainingDto.TrainingResponseDTO;
+import com.mariamkatamashvlii.gym.dto.trainingTypeDto.TrainingTypeDTO;
+import com.mariamkatamashvlii.gym.entity.Trainer;
+import com.mariamkatamashvlii.gym.entity.TrainingType;
+import com.mariamkatamashvlii.gym.entity.User;
+import com.mariamkatamashvlii.gym.exception.UserNotCreatedException;
+import com.mariamkatamashvlii.gym.generator.PasswordGenerator;
+import com.mariamkatamashvlii.gym.generator.UsernameGenerator;
+import com.mariamkatamashvlii.gym.repository.TrainerRepository;
+import com.mariamkatamashvlii.gym.repository.TrainingTypeRepository;
+import com.mariamkatamashvlii.gym.repository.UserRepository;
+import com.mariamkatamashvlii.gym.service.TrainerService;
+import jakarta.persistence.EntityNotFoundException;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Service;
+import org.springframework.web.server.ResponseStatusException;
+
+import java.time.LocalDate;
+import java.util.List;
+import java.util.UUID;
+
+@Slf4j
+@RequiredArgsConstructor
+@Service
+public class TrainerServiceImpl implements TrainerService {
+    private final TrainerRepository trainerRepo;
+    private final UserRepository userRepo;
+    private final TrainingTypeRepository trainingTypeRepo;
+    private final UsernameGenerator usernameGenerator;
+    private final PasswordGenerator passwordGenerator;
+
+    @Override
+    public RegistrationResponseDTO registerTrainer(RegistrationRequestDTO registrationRequestDTO) {
+        String transactionId = UUID.randomUUID().toString();
+        try {
+            log.info("[{}] Registering new trainer", transactionId);
+            User user = User.builder()
+                    .firstName(registrationRequestDTO.getFirstName())
+                    .lastName(registrationRequestDTO.getLastName())
+                    .username(usernameGenerator.generateUsername(registrationRequestDTO.getFirstName(), registrationRequestDTO.getLastName()))
+                    .password(passwordGenerator.generatePassword())
+                    .isActive(true)
+                    .build();
+            userRepo.save(user);
+            TrainingType type = trainingTypeRepo.findById(registrationRequestDTO.getSpecialization().getTrainingTypeId()).orElseThrow(() -> new EntityNotFoundException("TrainingType not found for id: " + registrationRequestDTO.getSpecialization().getTrainingTypeId()));
+            Trainer trainer = Trainer.builder()
+                    .specialization(type)
+                    .user(user)
+                    .build();
+            trainerRepo.save(trainer);
```

Comment:
there is a significant disadvantage here: two save operations. Instead, it's needed just to create Trainer BO and save it, inside there should be User object which will be creted by JPA automatically.